// halogenium/lib/Halogenium.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SyncLoader = exports.SquareLoader = exports.SkewLoader = exports.ScaleLoader = exports.RotateLoader = exports.RiseLoader = exports.RingLoader = exports.PulseLoader = exports.PacmanLoader = exports.MoonLoader = exports.GridLoader = exports.FadeLoader = exports.DotLoader = exports.ClipLoader = exports.BounceLoader = exports.BeatLoader = undefined;

var _BeatLoader2 = require('./BeatLoader');

var _BeatLoader3 = _interopRequireDefault(_BeatLoader2);

var _BounceLoader2 = require('./BounceLoader');

var _BounceLoader3 = _interopRequireDefault(_BounceLoader2);

var _ClipLoader2 = require('./ClipLoader');

var _ClipLoader3 = _interopRequireDefault(_ClipLoader2);

var _DotLoader2 = require('./DotLoader');

var _DotLoader3 = _interopRequireDefault(_DotLoader2);

var _FadeLoader2 = require('./FadeLoader');

var _FadeLoader3 = _interopRequireDefault(_FadeLoader2);

var _GridLoader2 = require('./GridLoader');

var _GridLoader3 = _interopRequireDefault(_GridLoader2);

var _MoonLoader2 = require('./MoonLoader');

var _MoonLoader3 = _interopRequireDefault(_MoonLoader2);

var _PacmanLoader2 = require('./PacmanLoader');

var _PacmanLoader3 = _interopRequireDefault(_PacmanLoader2);

var _PulseLoader2 = require('./PulseLoader');

var _PulseLoader3 = _interopRequireDefault(_PulseLoader2);

var _RingLoader2 = require('./RingLoader');

var _RingLoader3 = _interopRequireDefault(_RingLoader2);

var _RiseLoader2 = require('./RiseLoader');

var _RiseLoader3 = _interopRequireDefault(_RiseLoader2);

var _RotateLoader2 = require('./RotateLoader');

var _RotateLoader3 = _interopRequireDefault(_RotateLoader2);

var _ScaleLoader2 = require('./ScaleLoader');

var _ScaleLoader3 = _interopRequireDefault(_ScaleLoader2);

var _SkewLoader2 = require('./SkewLoader');

var _SkewLoader3 = _interopRequireDefault(_SkewLoader2);

var _SquareLoader2 = require('./SquareLoader');

var _SquareLoader3 = _interopRequireDefault(_SquareLoader2);

var _SyncLoader2 = require('./SyncLoader');

var _SyncLoader3 = _interopRequireDefault(_SyncLoader2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.BeatLoader = _BeatLoader3.default;
exports.BounceLoader = _BounceLoader3.default;
exports.ClipLoader = _ClipLoader3.default;
exports.DotLoader = _DotLoader3.default;
exports.FadeLoader = _FadeLoader3.default;
exports.GridLoader = _GridLoader3.default;
exports.MoonLoader = _MoonLoader3.default;
exports.PacmanLoader = _PacmanLoader3.default;
exports.PulseLoader = _PulseLoader3.default;
exports.RingLoader = _RingLoader3.default;
exports.RiseLoader = _RiseLoader3.default;
exports.RotateLoader = _RotateLoader3.default;
exports.ScaleLoader = _ScaleLoader3.default;
exports.SkewLoader = _SkewLoader3.default;
exports.SquareLoader = _SquareLoader3.default;
exports.SyncLoader = _SyncLoader3.default;
